zookeeper-server/src/main/java/org/apache/zookeeper/common/FileChangeWatcher.java

```diff
@@ -72,7 +71,7 @@ public FileChangeWatcher(Path dirPath, Consumer<WatchEvent<?>> callback) throws
 
         LOG.debug("Registering with watch service: {}", dirPath);
 
-        dirPath.register(watchService, new WatchEvent.Kind<?>[]{StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.OVERFLOW}, SensitivityWatchEventModifier.HIGH);
```

Comment:
Can't we detect the presence of that option at runtime and use it with reflection?

What's the impact of this patch?

Cc @enixon PTAL

Reply:
> Can't we detect the presence of that option at runtime and use it with reflection?

Code to do that would look similar to what is described in:
https://github.com/HotswapProjects/HotswapAgent/issues/41#issue-39256602

However, because of module access restrictions in future JVMs, there's a strong possibility that doing this would result in runtime exceptions in newer JVMs, unless users added the `jdk.unsupported` module to their module path at runtime... and even then, these unsupported APIs are supposed to be phased out entirely at some point... so there's a good chance this will add complexity which will never matter, because the complex code path will never be traversed (or will be traversed and cause an access restriction error at runtime).

> What's the impact of this patch?

That's the question I had, too (and the related: is it even worth it to try to keep the modifier?).

From what I could tell, the behavior this modifies is not well defined anyway. On some systems, the modifier *could* make file changes observed more quickly... but there's no guarantee of that... and there's no guarantee of seeing *every* change event on the file being watched, no matter what modifier is on it. Even with the modifier, it's probably still going to be on the order of a few seconds to notice changes. In my opinion, it's probably not worth it, but I'm not an expert on NIO stuff by far.

Reply:
I think this code was contributed from Facebook

@lvfangmin @enixon do you have any insight ?

Reply:
Looking into this a little bit more, it seems this utility is only ever used by the X509Util class to monitor changes to keyStore and trustStore files on disk, presumably for some sort of automatic reconfiguration.

The modifier seems to be an attempt to work around the slow polling implementation of the watch service on Mac OS X (https://bugs.openjdk.java.net/browse/JDK-7133447), which doesn't have a native implementation on that platform.

I can't imagine those files changing frequently enough for this to matter. But, even if they did change, I can't imagine that the optimization to automatically reconfigure a few seconds sooner matters either.

However, if automatic TLS reconfiguration was a critical need for somebody... there are lots of alternative watch service implementations out there to work around this issue. One is https://github.com/takari/directory-watcher ; I'm not advocating for using any of these... I don't know enough about them, but if it was important for somebody, they could consider contributing a change that makes use of one of those, instead of relying on the com.sun modifier to the built-in watch service that this PR removes.

Reply:
@eolivelli If @lvfangmin and @enixon do not reply, what happens next?

Reply:
Digging into this I also found a bunch of patch for OSX to speed up file change event.

There is not much info if this is still useful at all. In fact, most source (patch, stackoverflow etc.) I found are a few years old.

And in any case, I don't think there are many running production ZK on OSX. (If this change would primarily affect Linux systems I would be more keen on doing further research)

In short, LGTM